proto/Cargo.toml

```diff
@@ -12,6 +12,7 @@ rust-version = "1.73"
 [dependencies]
 hex = { version = "0.4" }
 miden-crypto = { workspace = true }
+miden-block-producer = { path = "../block-producer", version = "0.1"}
```

Comment:
Moved the structs to `miden-node-proto` [eea8581](https://github.com/0xPolygonMiden/miden-node/pull/55/commits/eea858141ac6516b037ab70434771516acdedfd9)

However, now that `miden-node-proto` contains domain types as well, we should probably rename it to `miden-node-objects`, or whatever else to capture the fact that we have both the protobuf definitions, as well as the "domain types" (i.e. the cleaned up types that we actually use in our code). I would probably remove [these re-exports](https://github.com/0xPolygonMiden/miden-node/blob/eea858141ac6516b037ab70434771516acdedfd9/proto/src/lib.rs#L7-L19) as well, since the `generated` types are no longer the focus of the crate.

Reply:
I'd like to think this through a bit more (and also curious what @hackaugusto thinks). My initial thinking was that `proto` is a crate for describing interfaces between different components rather than common structures. Maybe it's fine to have them be in the same place, but i could also imagine a separate `miden-node-core` crate which contains such common components.

Reply:
I think having shared structures here makes sense. But I don't understand why we have a copy of the protobuf struct in rust. For example:

```rust
pub struct AccountInputRecord {
    pub account_id: AccountId,
    pub account_hash: Digest,
    pub proof: MerklePath,
}

pub struct NullifierInputRecord {
    pub nullifier: Digest,
    pub proof: MerklePath,
}
```

Seems to be the same as:

```protobuf
// An account returned as a response to the GetBlockInputs
message AccountBlockInputRecord {
    account_id.AccountId account_id = 1;
    digest.Digest account_hash = 2;
    merkle.MerklePath proof = 3;
}

// A nullifier returned as a response to the GetBlockInputs
message NullifierBlockInputRecord {
    digest.Digest nullifier = 1;
    merkle.MerklePath proof = 3;
}
```

What are the benefits of this approach?

Reply:
The corresponding generated Rust code [has every field](https://github.com/0xPolygonMiden/miden-node/blob/50a6e2ed37d52596f5d84e06ede200d67070f505/proto/src/generated/responses.rs#L84-L91) as `Option<T>` by default for objects. This is a change from `proto2` to `proto3` syntax; see [Field Presence](https://protobuf.dev/programming-guides/field_presence/#background). So all we do with these domain types is ensure that every field is [indeed present](https://github.com/0xPolygonMiden/miden-node/blob/50a6e2ed37d52596f5d84e06ede200d67070f505/proto/src/conversion.rs#L259-L270).

Reply:
Oh, okay, so a bit of type safety. Makes sense. Next question would be: do we need that for every other crate? I suppose the endpoint is used basically only for `store -> txpool`, am I missing something?

Reply:
By `txpool`, I assume you mean [`block-producer`](https://github.com/0xPolygonMiden/miden-node/blob/8ca84b787d71e3cefbfa884e748f105ef844a62b/block-producer/Cargo.toml#L2)?

If I recall correctly, I think we wanted the `rpc` crate to also talk to the block producer through gRPC, so we would need similar type conversion stuff for `rpc -> txqueue`. Or basically anywhere else we use gRPC in general.

Reply:
> By txpool, I assume you mean [block-producer](https://github.com/0xPolygonMiden/miden-node/blob/8ca84b787d71e3cefbfa884e748f105ef844a62b/block-producer/Cargo.toml#L2)?

yes

> If I recall correctly, I think we wanted the rpc crate to also talk to the block producer through gRPC, so we would need similar type conversion stuff for rpc -> txqueue. Or basically anywhere else we use gRPC in general.

Not really. I created the `proto` crate exactly to avoid needing these conversions, the message type received by `rpc` can and IMO should be the same that is forwarded to `block-producer` to avoid needing so many conversions. (Unless, ofcourse, we are hydrating the data, but I don't this that would be the case here)

What I was trying to get at, is that if this is used by a single crate, than it should be defined there.